src/Protocol/ForgeHandshake.cpp

```diff
@@ -0,0 +1,365 @@
+
+// ForgeHandshake.cpp
+
+// Implements Forge protocol handshaking
+
+#include "Globals.h"
+#include "ForgeHandshake.h"
+#include "json/json.h"
+#include "../Server.h"
+#include "../Bindings/PluginManager.h"
+#include "../ClientHandle.h"
+#include "../Root.h"
+
+
+/** Discriminator byte values prefixing the FML|HS packets to determine their type. */
+enum
+{
+	Discriminator_ServerHello = 0,
+	Discriminator_ClientHello = 1,
+	Discriminator_ModList = 2,
+	Discriminator_RegistryData = 3,
+	Discriminator_HandshakeReset = -2,
+	Discriminator_HandshakeAck = -1,
+};
```

Comment:
Made this change, seems good. But for the client/server phases, I was surprised this fails to compile with a duplicate enum definition "COMPLETE":

```c++
/** Client handshake state phases. */
enum ClientPhase: unsigned char
{
	WAITINGSERVERDATA = 2,
	WAITINGSERVERCOMPLETE = 3,
	PENDINGCOMPLETE = 4,
	COMPLETE = 5,
};

/** Server handshake state phases. */
enum ServerPhase: unsigned char
{
	WAITINGCACK = 2,
	COMPLETE = 3,
};
```